homeassistant/components/weenect/__init__.py

```diff
@@ -0,0 +1,151 @@
+
+    session = async_get_clientsession(hass)
+    client = AioWeenect(username=username, password=password, session=session)
+
+    coordinator = WeenectDataUpdateCoordinator(hass, config_entry=entry, client=client)
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    for platform in PLATFORMS:
+        hass.async_add_job(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+
+    entry.add_update_listener(async_reload_entry)
+    return True
+
+
+class WeenectDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from the API."""
+
+    def __init__(
+        self, hass: HomeAssistant, config_entry: ConfigEntry, client: AioWeenect
+    ) -> None:
+        """Initialize."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=DEFAULT_UPDATE_RATE),
+        )
+        self.client = client
+        self.config_entry = config_entry
+        self.unsub_dispatchers: list[Callable[[], None]] = []
+        self.data: dict[int, Any] = {}
+
+    async def _async_update_data(self) -> dict[int, Any]:
+        """Update data via library."""
+        try:
+            data = await self.client.get_trackers()
+            data = self.transform_data(data)
+            self._detect_added_and_removed_trackers(data)
+            self._adjust_update_rate(data)
+            return data
+        except Exception as exception:
+            raise UpdateFailed(exception) from exception
+
+    def _detect_added_and_removed_trackers(self, data: dict[int, Any]) -> None:
+        """Detect if trackers were added or removed."""
+        assert self.config_entry is not None
+        added = data.keys() - self.data.keys()
+        async_dispatcher_send(
+            self.hass, f"{self.config_entry.entry_id}_{TRACKER_ADDED}", added
+        )
+        removed = self.data.keys() - data.keys()
+        async_dispatcher_send(
+            self.hass, f"{self.config_entry.entry_id}_{TRACKER_REMOVED}", removed
+        )
+
+    def _adjust_update_rate(self, data: dict[int, Any]) -> None:
+        """Set the update rate to the shortest update rate of all trackers."""
+        update_rate = timedelta(seconds=DEFAULT_UPDATE_RATE)
+        for tracker in data.values():
+            tracker_rate = self.parse_duration(tracker["last_freq_mode"])
+            if tracker_rate and tracker_rate < update_rate:
+                update_rate = tracker_rate
+        self.update_interval = update_rate
+        _LOGGER.debug("Setting update_interval to %s", update_rate)
+
+    @staticmethod
+    def transform_data(data: Any) -> dict[int, Any]:
+        """Extract trackers from list and put them in a dict by tracker id."""
+        result = {}
+        for tracker in data["items"]:
+            result[tracker["id"]] = tracker
+        return result
```

Comment:
More succinct (and a tiny bit faster):
```suggestion
        return {tracker["id"]: tracker for tracker in data["items"]}
```